go/mysqlconn/auth_server.go

```diff
@@ -0,0 +1,111 @@
+package mysqlconn
+
+import (
+	"crypto/rand"
+	"crypto/sha1"
+
+	log "github.com/golang/glog"
+)
+
+// AuthServer is the interface that servers must implement to validate
+// users and passwords. It has two modes:
+//
+// 1. using salt the way MySQL native auth does it. In that case, the
+// password is not sent in the clear, but the salt is used to hash the
+// password both on the client and server side, and the result is sent
+// and compared.
+//
+// 2. sending the user / password in the clear (using MySQL Cleartext
+// method). The server then gets access to both user and password, and
+// can authenticate using any method. If SSL is not used, it means the
+// password is sent in the clear. That may not be suitable for some
+// use cases.
+type AuthServer interface {
```

Comment:
I was thinking Salt() would do it. When Salt() is called, you can connect to the MySQL backend, get its salt, return that, and when ValidateHash is called, call the MySQL backend again (you then need to have a salt -> backend map, but that's easy).